### PR TITLE
Implement RFC 640

### PR DESCRIPTION
**Pull request by sfackler**
I've made some minor changes from the implementation attached to the RFC to try to minimize codegen. The methods now take `&Debug` trait objects rather than being parameterized and there are inlined stub methods that call to non-inlined methods to do the work.

r? @alexcrichton 

cc @huonw for the `derive(Debug)` changes.

### CODE REVIEW

**Comment by brson** [comment]
re https://github.com/rust-lang/rfcs/blob/master/text/0640-debug-improvements.md

**Comment by alexcrichton** [comment]
I'll take a closer look at this tomorrow, but I'm curious why you applied `#[inline]` attributes at all? It looks like none of these methods require inlining (e.g. none have type parameters) and if you avoid the `#[inline]` attribute entirely then they'll all just naturally not be inlined because they're across the crate boundary.

**Comment by sfackler** [comment]
The public API takes `self` by value so that we can return the `fmt::Result` at the call to `finish`. But if we didn't inline it at all there'd be overhead from repeatedly copying the builder struct around. The current setup has inlineable stub functions that take `self` by value and do nothing but call a private non-inlineable method that takes `&mut self` and then returns `self`. The idea is that the end result is as if the API took `&mut self` at the assembly level while behaving as if it takes `self`.

**Comment by alexcrichton** [comment]
> But if we didn't inline it at all there'd be overhead from repeatedly copying the builder struct around.

Ah ok. If we're trying to minimize codegen though I would probably recommend a `&mut self`-based builder as it would involve far fewer moves. The `finish` method could replace the inner result with `Ok(())`, for example, so it doesn't necessarily have to consume.

In general I'm not so sure about the application of `#[inline]` throughout these methods. I would be much more convinced with some data though! Without some statistics in favor of the `#[inline]`, I think I'd personally prefer they'd all be removed for now. One interesting statistic would be the before/after size of librustdoc. I know that the lib has probably hundreds of `#[derive(Debug)]` and if this cuts down on codegen it should really help that size.

Otherwise, my only other concern is the usage in `#[derive]`. I'm a little unsure about how we lint stability in `#[derive]`-based implementations, but I just wanted to confirm that this does not require the `core` feature gate? I personally don't necessarily like the compiler being able to call unstable internals (e.g. the expanded AST won't actually compiled b/c it calls unstable code) where other code cannot, but it's also something we're at liberty to change so I'd be fine giving this a whirl to see what happens!

**Comment by sfackler** [comment]
I will collect some statistics.

The deriving implementation uses the functionality @huonw added to be able to call unstable APIs without triggering feature gates.

**Comment by alexcrichton** [comment]
Ah excellent. cc @huonw on the `#[derive]` changes then, but I think they're all pretty ok

File: src/libsyntax/ext/deriving/show.rs [hunk 1]
```diff
@@ -72,70 +67,53 @@ fn show_substructure(cx: &mut ExtCtxt, span: Span,
         }
     };
 
-    let mut format_string = String::from_str(&token::get_ident(name));
-    // the internal fields we're actually formatting
-    let mut exprs = Vec::new();
+    // We want to make sure we have the expn_id set so that we can use unstable methods
+    let span = Span { expn_id: cx.backtrace(), .. span };
```
Comment (huonw) [inline src/libsyntax/ext/deriving/show.rs]: Did you find that this was necessary? I'm under the impression that the macro infrastructure manages re-spanning everything after a macro expansion?

**Comment by huonw** [comment]
Looks great. Is it worth putting these under a different feature gate to `core`?

File: src/libsyntax/ext/deriving/show.rs (hunk 1 above)
Comment (sfackler) [inline src/libsyntax/ext/deriving/show.rs]: It did not appear to respan - I think only the `macro_rules!` infrastructure may do that.

**Comment by sfackler** [comment]
So it turns out that no-inline silliness actually ends up a tiny bit smaller than inline silliness: https://gist.github.com/sfackler/a9300bce13ad166cd35b. Looks like the builder structs are small enough that the copies don't matter.

I pushed a commit that cuts out all the inline annotations.

**Comment by sfackler** [comment]
Oh, and there doesn't appear to be much of a difference one way or another for overall binary sizes before and after the `derive(Debug)` change.

**Comment by alexcrichton** [comment]
> So it turns out that no-inline silliness actually ends up a tiny bit smaller than inline silliness

Thanks for investigating!

> Oh, and there doesn't appear to be much of a difference one way or another for overall binary sizes before and after the derive(Debug) change.

Ah well, it probably at least paves the way in the future for such changes though.

**Comment by alexcrichton** [comment]
This looks great to me! r=me with a feature name that's not `core` (e.g. `debug_builders` or something like that)

**Comment by sfackler** [comment]
@bors retry

Seems like a spurious error??

**Comment by SimonSapin** [comment]
Should `assert_eq!` default to using this? (I just found out about this, it looks great!)

**Comment by liigo** [comment]
keywords: pretty debug builders
https://github.com/rust-lang/rfcs/pull/640
(to help people find this rfc implementation)

**Comment by sunjay** [comment]
@SimonSapin It would be great is `assert_eq!` used this. I don't know what a good way to integrate it in my tests right now without that.

**Pull request merged**